### ISSUE TITLE
One way transmission links broken when used in both directions

### ISSUE THREAD

**Issue by timtroendle** [calliope-project/calliope#215]
## Problem description

I want to model net transfer capacities which are direction dependent. For example, I may want to model that from SWE to NOR the capacity is 3, but from NOR to SWE it is 4. The documentation doesn't say this is possible, but I assumed I could build one one-way transmission link each.

While Calliope does not complain, I believe the capacities tend to be the same in the end. In the example below they are both 3, but in my real model they were "4" (the higher value).

If this is not a feature, Calliope should complain about the second link one-way link to warn the user that this may not lead to the expected result.

## Steps to reproduce the problem

Minimal model

```yaml
model:
    name: 'test model'
    calliope_version: 0.6.4
    timeseries_data_path: './'
    subset_time: ['2015-01-01', '2015-01-01']
run:
    solver: gurobi
    solver_io: python
    mode: plan
techs:
    demand_elec:
        essentials:
            name: 'Electricity demand'
            parent: demand
            carrier: electricity
        constraints:
            resource: file=electricity-demand.csv
    simple_supply:
        essentials:
            parent: supply
            carrier: electricity
        constraints:
            lifetime: 25
            energy_cap_max: .inf
        costs:
            monetary:
                interest_rate: 0.1
                energy_cap: 10
    ac_transmission:
        essentials:
            name: "High voltage AC transmission line"
            parent: transmission
            carrier: electricity
locations:
    SWE,NOR:
        techs: {demand_elec, simple_supply}
links:
    SWE,NOR.techs:
        ac_transmission:
            constraints:
                energy_cap_equals: 3
                one_way: true
    NOR,SWE.techs:
        ac_transmission:
            constraints:
                energy_cap_equals: 4
                one_way: true
```

And here is why I assume it's broken:

```python
In [6]: model = calliope.Model("build/model/minimal_model.yaml")

In [7]: model.run()

In [8]: model.get_formatted_array("energy_cap_equals").to_dataframe().loc["NOR"].loc["ac_transmission:SWE"]
Out[8]:
energy_cap_equals    3.0
Name: ac_transmission:SWE, dtype: float64

In [9]: model.get_formatted_array("energy_cap_equals").to_dataframe().loc["SWE"].loc["ac_transmission:NOR"]
Out[9]:
energy_cap_equals    3.0
Name: ac_transmission:NOR, dtype: float64

In [10]: model.get_formatted_array("energy_cap").to_dataframe().loc["SWE"].loc["ac_transmission:NOR"]
Out[10]:
energy_cap    3.0
Name: ac_transmission:NOR, dtype: float64

In [11]: model.get_formatted_array("energy_cap").to_dataframe().loc["NOR"].loc["ac_transmission:SWE"]
Out[11]:
energy_cap    3.0
Name: ac_transmission:SWE, dtype: float64
```

## Calliope version

0.6.4-dev

**Comment by timtroendle** [comment]
@sjpfenninger could you have a quick look and tell me whether I'm right in my assumption that this does not work as I expected it to work?

**Comment by brynpickering** [comment]
I'll admit, I'm surprised it doesn't break during the optimisation... There is a constraint to directly connect two directions of a link and force their capacities to be the same:

```
[in] model._backend_model.symmetric_transmission_constraint.pprint()

[out] symmetric_transmission_constraint : Size=2, Index=loc_techs_symmetric_transmission_constraint, Active=True
    Key                      : Lower : Body                                                                        : Upper : Active
    NOR::ac_transmission:SWE :   0.0 : energy_cap[NOR::ac_transmission:SWE] - energy_cap[SWE::ac_transmission:NOR] :   0.0 :   True
    SWE::ac_transmission:NOR :   0.0 : energy_cap[SWE::ac_transmission:NOR] - energy_cap[NOR::ac_transmission:SWE] :   0.0 :   True

```

But this contradicts the  `energy_cap_equals` constraints:

```
[in] model._backend_model.energy_capacity_constraint.pprint()

[out] energy_capacity_constraint : Size=2, Index=loc_techs_energy_capacity_constraint, Active=True
    Key                      : Lower                                       : Body                                 : Upper                                       : Active
    NOR::ac_transmission:SWE : energy_cap_equals[NOR::ac_transmission:SWE] : energy_cap[NOR::ac_transmission:SWE] : energy_cap_equals[NOR::ac_transmission:SWE] :   True
    SWE::ac_transmission:NOR : energy_cap_equals[SWE::ac_transmission:NOR] : energy_cap[SWE::ac_transmission:NOR] : energy_cap_equals[SWE::ac_transmission:NOR] :   True
```

**Comment by brynpickering** [comment]
@timtroendle Having looked into this, I think allowing the same two locations to be linked by different links (for the same technology) will be a bit difficult, because:

1. directly connecting links occurs in two ways:
a. By linking the energy capacity in `symmetric_transmission_constraint`.
b. By linking the energy production it one location with the energy consumption at at the location at the other end of the link.
a. would be easy to remove, but b. would still effectively limit the link. `carrier_prod` and `carrier_con` are currently limited by the technology energy capacity, so the limit on `carrier_con` of one direction of the link would limit the `carrier_prod` of the link in the other direction. 

2. Costs are currently based on the link having two `loc_techs` associated with it. I.e. `loc1::tech:loc2` and `loc2::tech:loc1`. As such, the investment costs are always halved for transmission techs, with each direction of the link costing half the total cost of the link. If the links were separated, this halving would still take place, but I assume you would want each direction to be costed separately. 

So, I'm not sure there is an easy solution to having transmission techs being mutually exclusive in each direction, based on their current implementation in the backend. 

Would it be easy enough to have two different technologies to describe each independent direction, or is that more hassle than trying to reform our current implementation?

p.s. I see now why the model doesn't break on having contradictory constraints. In the YAML preprocessing, somewhere as part of building `model_run`, `energy_cap_equals` ends up being 3 in both directions (see `model._model_run.locations.NOR.links` and `model._model_run.locations.SWE`). You can make it equal 4 for both of them by just switching the constraint between the two links. More specifically, it is caused by lines 229 - 237 in `calliope/core/preprocess/locations.py`:

```yaml
processed_links.set_key(
    '{}.links.{}.techs.{}'.format(loc_from, loc_to, tech_name),
    tech_settings.copy()
)

processed_links.set_key(
    '{}.links.{}.techs.{}'.format(loc_to, loc_from, tech_name),
    tech_settings.copy()
)
```

For the same link defined twice, once in each direction, the direction which is processed last will have its settings take precedence. Just shows how little we expected a user to ever define a link twice with different data! Definitely this current behaviour should raise a warning/exception.

**Comment by timtroendle** [comment]
I think this should raise rather an exception than a warning.

Did I understand correctly: a pragmatic workaround is to have two different technologies? For my current needs that'll be fine. And I wonder whether this couldn't be the official way to do it. Using tech groups this can be done easily.

**Comment by brynpickering** [comment]
> Did I understand correctly: a pragmatic workaround is to have two different technologies? For my current needs that'll be fine. And I wonder whether this couldn't be the official way to do it. Using tech groups this can be done easily.

Yep, with your example, something like this should do it:

```yaml
model:
    name: 'test model'
    calliope_version: 0.6.4
    timeseries_data_path: './'
    subset_time: ['2005-01-01', '2005-01-01']
run:
    solver: gurobi
    solver_io: python
    mode: plan
    
tech_groups:
    ac_transmission:
        essentials:
            parent: transmission
            carrier: electricity

techs:
    demand_elec:
        essentials:
            name: 'Electricity demand'
            parent: demand
            carrier: electricity
        constraints:
            resource: file=electricity-demand.csv
    simple_supply:
        essentials:
            parent: supply
            carrier: electricity
        constraints:
            lifetime: 25
            energy_cap_max: .inf
        costs:
            monetary:
                interest_rate: 0.1
                energy_cap: 10
    ac_transmission_nor_swe:
        essentials:
            parent: ac_transmission
            name: 'NOR -> SWE link'
        constraints:
            energy_cap_equals: 3
            one_way: true
            
    ac_transmission_swe_nor:
        essentials:
            parent: ac_transmission
            name: 'SWE -> NOR link'
        constraints:
            energy_cap_equals: 4
            one_way: true
locations:
    SWE,NOR:
        techs: {demand_elec, simple_supply}
links:
    SWE,NOR.techs:
        ac_transmission_swe_nor:
    NOR,SWE.techs:
        ac_transmission_nor_swe:
```

**Comment by sjpfenninger** [comment]
I agree:

1. @brynpickering's proposed workaround is the way to go for now,
2. we definitely need to raise an exception if the same transmission tech is defined in both different directions,
3. and possibly re-think how transmission techs are defined for a future version of Calliope (but not for 0.6.4).

I can address number 2 at some point next week.

**Comment by timtroendle** [comment]
Thanks both. Just to be sure, the two tranmission techs don't have to be location specific, do they? So the following should work as well:

```yaml
tech_groups:
    ac_transmission:
        essentials:
            parent: transmission
            carrier: electricity

techs:
    uni_directional_ac_transmission_A:
        essentials:
            parent: ac_transmission
            name: 'Uni directional AC transmission link A.'
        constraints:
            one_way: true
            
    uni_directional_ac_transmission_B:
        essentials:
            parent: ac_transmission
            name: 'Uni directional AC transmission link B.'
        constraints:
            one_way: true
links:
    SWE,NOR.techs:
        uni_directional_ac_transmission_A:
             energy_cap_equals: 4
    NOR,SWE.techs:
        uni_directional_ac_transmission_B:
             energy_cap_equals: 3
```

**Comment by sjpfenninger** [comment]
@timtroendle Yes, that should work

**Comment by timtroendle** [comment]
Another thing I just figured: the transmission plot seems to not be able to handle uni directional links. Using current master, this works:

```Python
import calliope
calliope.examples.national_scale().plot.transmission()
```

but this fails:

```Python
import calliope
model = calliope.read_netcdf("my-model-results.nc")
model.plot.transmission()
```

And here is the error message:
```Python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-517ea42645aa> in <module>
----> 1 model.plot.transmission()

~/Develop/ose-model-comparison/calliope/calliope/analysis/plotting/plotting.py in transmission(self, **kwargs)
    188     def transmission(self, **kwargs):
    189         self.check_optimality()
--> 190         data, layout = plot_transmission(self._model, **kwargs)
    191         return _plot(data, layout, **kwargs)
    192

~/Develop/ose-model-comparison/calliope/calliope/analysis/plotting/transmission.py in plot_transmission(model, mapbox_access_token, **kwargs)
    139
    140     if hasattr(model, 'results'):
--> 141         energy_cap = _get_data(model, 'energy_cap')
    142         carrier_prod = _get_data(model, 'carrier_prod', sum_dims=['timesteps', 'carriers'])
    143         carrier_con = _get_data(model, 'carrier_con', sum_dims=['timesteps', 'carriers'])

~/Develop/ose-model-comparison/calliope/calliope/analysis/plotting/transmission.py in _get_data(model, var, sum_dims)
     55         var_da = var_da.sum(dim=sum_dims)
     56     techs = list(
---> 57         set(model._model_data.techs_transmission.values).intersection(var_da.techs.values)
     58     )
     59     var_df = var_da.loc[dict(techs=techs)].to_pandas()

~/miniconda3/envs/ose-model-comparison/lib/python3.6/site-packages/xarray/core/common.py in __getattr__(self, name)
    175                     return source[name]
    176         raise AttributeError("%r object has no attribute %r" %
--> 177                              (type(self).__name__, name))
    178
    179     def __setattr__(self, name, value):

AttributeError: 'Dataset' object has no attribute 'techs_transmission'
```

**Comment by brynpickering** [comment]
Are you sure this is related to one way transmission, and not loading a model from file before plotting transmission?

**Comment by timtroendle** [comment]
No, wasn't sure so I gave it a try just now. This works fine:

```Python
import calliope
model = calliope.examples.national_scale()
model.run()
model.to_netcdf("test.nc")
calliope.read_netcdf("test.nc").plot.transmission()
```

So it's not the loading a model from file part.

**Comment by sjpfenninger** [comment]
#223 has addressed this issue for now - #234 tracks the issue with plotting and #166 with changes to transmission definitions more generally.
